web/interceptor/access_control.go

```diff
@@ -0,0 +1,167 @@
+	"google.golang.org/grpc"
+)
+
+type (
+	role      int
+	roles     []role
+	requestID interface {
+		IDFor(string) uint64
+	}
+)
+
+const (
+	none role = iota
+	// user role implies that user attempts to access information about himself.
+	user
+	// group role implies that the user is a course student + a member of the given group.
+	group
+	// student role implies that the user is enrolled in the course with any role.
+	student
+	// teacher: user enrolled in the course with teacher status.
+	teacher
+	// courseAdmin: an admin user who is also enrolled into the course.
+	courseAdmin
+	// admin is the user with admin privileges.
+	admin
+)
+
+// If there are several roles that can call a method, a role with the least privilege must come first.
+var accessRolesFor = map[string]roles{
+	"GetUser":                 {none},
+	"GetCourse":               {none},
+	"GetCourses":              {none},
+	"CreateEnrollment":        {user},
+	"UpdateCourseVisibility":  {user},
+	"GetCoursesByUser":        {user},
+	"UpdateUser":              {user, admin},
+	"GetEnrollmentsByUser":    {user, admin},
+	"GetSubmissions":          {student, group, teacher, courseAdmin},
+	"GetGroupByUserAndCourse": {student, teacher},
+	"CreateGroup":             {group, teacher},
+	"GetGroup":                {group, teacher},
+	"GetAssignments":          {student, teacher},
+	"GetEnrollmentsByCourse":  {student, teacher},
+	"GetRepositories":         {student, teacher},
+	"UpdateGroup":             {teacher},
+	"DeleteGroup":             {teacher},
+	"GetGroupsByCourse":       {teacher},
+	"UpdateCourse":            {teacher},
+	"UpdateEnrollments":       {teacher},
+	"UpdateAssignments":       {teacher},
+	"UpdateSubmission":        {teacher},
+	"UpdateSubmissions":       {teacher},
+	"RebuildSubmissions":      {teacher},
+	"CreateBenchmark":         {teacher},
+	"UpdateBenchmark":         {teacher},
+	"DeleteBenchmark":         {teacher},
+	"CreateCriterion":         {teacher},
+	"UpdateCriterion":         {teacher},
+	"DeleteCriterion":         {teacher},
+	"CreateReview":            {teacher},
+	"UpdateReview":            {teacher},
+	"GetReviewers":            {teacher},
+	"IsEmptyRepo":             {teacher},
+	"GetSubmissionsByCourse":  {courseAdmin},
+	"GetUserByCourse":         {teacher, admin},
+	"GetUsers":                {admin},
+	"GetOrganization":         {admin},
+	"CreateCourse":            {admin},
+}
+
+// AccessControl checks user information stored in the JWT claims against the list of roles required to call the method.
+func AccessControl(logger *zap.SugaredLogger, tm *auth.TokenManager) grpc.UnaryServerInterceptor {
+	return func(ctx context.Context, request interface{}, info *grpc.UnaryServerInfo, handler grpc.UnaryHandler) (interface{}, error) {
+		method := info.FullMethod[strings.LastIndex(info.FullMethod, "/")+1:]
+		req, ok := request.(requestID)
+		// The GetUserByCourse method sends a CourseUserRequest which has no IDs and needs a database query.
```

Comment:
Is this comment still relevant here? It feels out of place.

Reply:
You are right. Removing.